**ML1-Notebook4.1**

***Author: Açelya Yıldırım***
***Date: 10.10.2025***

# Metrics

We have used simple functions in the previous assignments, such as MSE in regression problems or accuracy in classification problems, to quantify the goodness-of-fit of the model. While in regression problems the functions are based on an error calculated in one way or another (mean error, mean square error, etc.), in classification problems other types of metrics can be derived depending on what the problem being solved is like. Many of these metrics, at least those that will be used in practice, are based on the calculation of the confusion matrix.

A confusion matrix is a square matrix, with as many rows and columns as classes, showing the distribution of patterns in classes, and the classification performed by the model. Usually the rows show how the model has performed the classification, and the columns show the actual classification values, although this may vary depending on the source consulted.

The simplest case corresponds to 2 classes, where one is considered "negative" and the other "positive". A two-class confusion matrix would be as follows:

| |                      | **Predicted**          |                        |
|--------|----------------------|------------------------|------------------------|
|  |                      | **Negative**           | **Positive**           |
| **Real** | **Negative**  | TN (True Negative)     | FP (False Positive)    |
|  | **Positive**  | FN (False Negative)    | TP (True Positive)     |

This confusion matrix contains 4 values, which can be divided

- according to the output of the model: positive or negative.
- depending on whether the model is wrong or not: true or false. 

Thus, these 4 values are called true negatives (TN), false positives (FP), false negatives (FN) and true positives (TP). For example, false negatives would be the number of patterns that the system has classified as negative, and has been wrong because they were actually positive.

From this confusion matrix, different metrics can be calculated. Depending on the problem you are working on, it will be more interesting to follow one or the other. Some of the most most commonly used metrics are:

- **Accuracy**. Ratio of patterns in which the prediction is correct. Calculated as $$\frac{TN+TP}{TN+TP+FN+FP}$$
- **Error rate**. Ratio of patterns in which the prediction is wrong. Calculated as $$\frac{FP+FN}{TN+TP+FN+FP}$$
- **Sensitivity** or **recall**. Indicates the probability that a positive classification result is obtained for a positive case. It is calculated as $$\frac{TP}{FN+TP}$$
    - In a medical test, the test sensitivity represents the probability that a sick (positive) subject will have a positive test result.
- **Specificity**. Indicates the probability that a negative classification result is obatined for a negative case. It is calculated as $$\frac{TN}{FP+TN}$$
    - The specificity of a test represents the probability that a healthy (negative) subject will have a negative test result.
- **Precision** or **positive predictive value**. Ratio of positive patterns that have been correctly classified. Calculated as $$\frac{TP}{TP+FP}$$
- **Negative predictive value**. Ratio of positive patterns that have been correctly classified. Calculated as $$\frac{TN}{TN+FN}$$
- **F-score**, **F1-score** or **F-measurement**. It is defined as the harmonic mean of precision and recall.

It is worth clarifying that these metrics, as well as others seen in theory class (ROC curve, Kappa index) are used to assess already trained classifiers, not to perform the training process. To be trained, each model has its own function to quantify the error or goodness of fit, such as the cross-entropy function in the case of neural networks.

The accuracy is probably the most commonly used value, as it indicates the success rate of the classifier in a simple way. However, depending on the problem you are working with, it may not be the most appropriate metric. For example, in a mass population-based test for a disease where it is known that most people do not have the disease, a model that classifies everyone as negative (healthy) will have a very high accuracy, even though the model does not actually do anything.

For this reason, it is necessary to assess which metric or metrics are the most used for each kind of problem. In many problems where the different classes are of equal importance, the accuracy value may be enough. However, in other problems, it may be of more interest to evaluate the situations in which a positive response is or should be produced by the model, as it could indicate something critical, such as detecting a disease or raising some kind of alarm. For this reason, sensitivity and positive predictive value values are often taken into account in addition to accuracy. There is a more extensive discussion of this in the theory notes, but a possible informal guide might be the following:

- If you want to minimise the number of positives incorrectly classified as negative (e.g. maximise the number of correctly diagnosed sick subjects, or maximise the number of alarms given for risky situations), the indicated metric is sensitivity (recall).
- If one wishes to minimise the number of samples incorrectly classified as positives (false positives, e.g. sick subjects diagnosed as healthy, or situations where an alarm should not has been raised but it was), the indicated metric is the positive predictive value (precision).

Therefore, the most appropriate metric depends entirely on the specific problem, according to the relative importance of the classifier output and its behaviour. In this type of problem, the F-score is a metric that may be more useful than accuracy.

Another issue to be considered is the data imbalance. Accuracy is a metric that gives a "global" view, which can be misleading when the distribution in classes is unbalanced. In these cases, F-score is a better metric. Having unbalanced databases is very common, which provides an extra argument for using F-score rather than accuracy.

Finally, if you have more than two classes, it is possible to build a confusion matrix in a similar way by having one row and column per class. For example:


In these cases, it is no longer possible to speak of positive or negative patterns, since there are more than two classes, nor to take values for sensitivity or positive predictive value. However, this confusion matrix can offer very interesting information when it comes to understanding how the model works, assessing which are the classes between which the model finds it easiest and most difficult to separate.

### Question 4.1
> ❓ If the pattern set has been divided into training and test subsets, which subset should be used to calculated the confusion matrix?

The confusion matrix should be calculated using the test subset. This is because the confusion matrix is used to evaluate the model's performance on unseen data, not on the data it was trained on. If we calculate the confusion matrix on the training set, we cannot detect overfitting, as the model may have memorized the training patterns and show artificially high accuracy. The true performance of a model can only be measured on data it has never seen before, which is represented by the test set. This gives us a realistic assessment of how the model will perform in real-world scenarios with new, unseen patterns.

In this assignment, you are asked to:

1. Develop a function called `confusionMatrix` which takes two vectors of equal length (the number of patterns), the first one containing the outputs obtained by a model (`outputs`) and the second with the desired outputs (`targets`), both of type `AbstractArray{Bool,1}`. This function should return:
    - Accuracy
    - Error rate
    - Sensitivity
    - Specificity
    - Positive predicitive value
    - Negative predicitve value
    - F-score
    - Confusion matrix, as an object of type `Array{Int64,2}` with two rows and two columns
    
    As this function is being fed with boolean-valued vectors, it will be applicable to problems with two classes (positive and negative cases).

    It is necessary to consider some particular situations when calculated the required classification metrics.
    
        - If every pattern is a true negative, neither the sensitivity nor the positive predictive value can be calculated. In this case the system works correctly, so these two metrics will be 1.
        - Similarly, neither the specificity nor the negative predictive value can be obtained and if every pattern is a true positive, so both metrics have to be manually set to 1.
        - If neither of these two cases has occurred and there is still any metric which cannot be calculated, it will take the value of 0. For example, if the sensitivity could not be calculated, it means that there was no pattern with a positive desired output.
        - It both sensitivity and positive predictive values are equal to 0, the value of F-score cannot be obtained, and thus it will be 0.
        
    Do not use loops inside the developed function.


In [4]:
function confusionMatrix(outputs::AbstractArray{Bool,1}, targets::AbstractArray{Bool,1})
    # Step 1: Calculate TP, TN, FP, FN 
    TP = sum(outputs .& targets)
    TN = sum(.!outputs .& .!targets)
    FP = sum(outputs .& .!targets)
    FN = sum(.!outputs .& targets)
    
    # Step 2: Check exceptions and calculate metrics
    
    # Total number of patterns
    total = length(outputs)
    
    # Accuracy and Error Rate
    accuracy = (TP + TN) / total
    errorRate = (FP + FN) / total
    
    # Sensitivity (Recall)
    if (TP + FN) == 0  # No positive patterns
        sensitivity = (TN == total) ? 1.0 : 0.0  # 1 if all are TN, 0 otherwise
    else
        sensitivity = TP / (TP + FN)
    end
    
    # Specificity
    if (TN + FP) == 0  # No negative patterns
        specificity = (TP == total) ? 1.0 : 0.0  # 1 if all are TP, 0 otherwise
    else
        specificity = TN / (TN + FP)
    end
    
    # Positive Predictive Value (Precision)
    if (TP + FP) == 0
        ppv = (TN == total) ? 1.0 : 0.0
    else
        ppv = TP / (TP + FP)
    end
    
    # Negative Predictive Value
    if (TN + FN) == 0
        npv = (TP == total) ? 1.0 : 0.0
    else
        npv = TN / (TN + FN)
    end
    
    # F-score
    if sensitivity == 0.0 && ppv == 0.0
        fscore = 0.0
    else
        fscore = 2 * (ppv * sensitivity) / (ppv + sensitivity)
    end
    
    # Confusion Matrix
    confMatrix = [TN FP; FN TP]
    
    # Return tuple
    return accuracy, errorRate, sensitivity, specificity, ppv, npv, fscore, confMatrix
end

confusionMatrix (generic function with 1 method)

In [6]:
# Test
outputs_test = [true, false, true, true, false]
targets_test = [true, false, false, true, true]

result = confusionMatrix(outputs_test, targets_test)
println(result)

(0.6, 0.4, 0.6666666666666666, 0.5, 0.6666666666666666, 0.5, 0.6666666666666666, [1 1; 1 2])


The first confusionMatrix function was implemented, which takes boolean vectors as inputs. The function calculates all required metrics without using any loops, utilizing vectorized operations instead. When tested with sample data, the function correctly computed the confusion matrix showing 2 true positives, 1 true negative, 1 false positive, and 1 false negative. The resulting metrics indicated 60% accuracy with a sensitivity and positive predictive value of 66.7%, and specificity and negative predictive value of 50%. The F-score was calculated as 0.667, demonstrating a balanced performance between precision and recall. All edge cases were properly handled, including scenarios where all patterns might be true negatives or true positives, ensuring robust metric calculations even in extreme situations.

2. Many models (e.g. artificial neural networks) do not return a categorial output for each pattern, but the probability that it is "positive". For this reason, it is requested to develop a function with the same name as the previous one, whose first parameter is not a vector of boolean values but a vector of real values (of type `AbstractArray{<:Real}`). It also receives an optional third parameter with a threshold, with a default value, which is used to apply the previous function and return, therefore, the same values.

In [8]:
function confusionMatrix(outputs::AbstractArray{<:Real,1}, targets::AbstractArray{Bool,1}; threshold::Real=0.5)
    # Convert Real outputs to Boolean using threshold
    boolOutputs = outputs .>= threshold
    
    # Call the first confusionMatrix function
    return confusionMatrix(boolOutputs, targets)
end

confusionMatrix (generic function with 2 methods)

In [10]:
# Test the second confusionMatrix function with Real outputs
outputs_real = [0.1, 0.8, 0.3, 0.9, 0.2]
targets_test = [true, false, false, true, true]

# Call the function with default threshold (0.5)
result_real = confusionMatrix(outputs_real, targets_test)
println("Result with default threshold (0.5): ", result_real)

# Test with different threshold
result_real_03 = confusionMatrix(outputs_real, targets_test, threshold=0.3)
println("Result with threshold 0.3: ", result_real_03)

Result with default threshold (0.5): (0.4, 0.6, 0.3333333333333333, 0.5, 0.5, 0.3333333333333333, 0.4, [1 1; 2 1])
Result with threshold 0.3: (0.2, 0.8, 0.3333333333333333, 0.0, 0.3333333333333333, 0.0, 0.3333333333333333, [0 2; 2 1])


The second confusionMatrix function handles real-valued outputs by converting them to boolean values using a specified threshold. Testing with real outputs [0.1, 0.8, 0.3, 0.9, 0.2] demonstrated that the threshold parameter significantly impacts classification results. With the default threshold of 0.5, the function achieved 40% accuracy, while lowering the threshold to 0.3 reduced accuracy to 20% due to an increase in false positives. This behavior confirms that the function correctly applies the threshold comparison and properly delegates to the boolean version of confusionMatrix, maintaining consistency across different threshold values.

3. Develop two functions with the same name, `printConfusionMatrix`, that receive the model outputs and the desired outputs, call the previous functions and display the results obtained, including the confusion matrix. One of these functions shall receive a vector of model classifications (`outputs`) of type `AbstractArray{Bool,1}`, while for the other one this parameter shall be a vector of real values (of type `AbstractArray{<:Real}`). These functions will make calls to the previous functions.

In [12]:
function printConfusionMatrix(outputs::AbstractArray{Bool,1}, targets::AbstractArray{Bool,1})
    # Call confusionMatrix function
    acc, errorRate, sens, spec, ppv, npv, fscore, confMat = confusionMatrix(outputs, targets)
    
    # Print results
    println("=== Confusion Matrix Results ===")
    println("Accuracy: ", acc)
    println("Error Rate: ", errorRate)
    println("Sensitivity (Recall): ", sens)
    println("Specificity: ", spec)
    println("Positive Predictive Value (Precision): ", ppv)
    println("Negative Predictive Value: ", npv)
    println("F-score: ", fscore)
    println("\nConfusion Matrix:")
    println("       Predicted")
    println("       Neg  Pos")
    println("Real Neg ", confMat[1,1], "    ", confMat[1,2])
    println("Real Pos ", confMat[2,1], "    ", confMat[2,2])
    println("================================")
end

printConfusionMatrix (generic function with 1 method)

In [14]:
# Test the print functions
outputs_bool = [true, false, true, true, false]
targets_bool = [true, false, false, true, true]

# Call the function and show the prints
printConfusionMatrix(outputs_bool, targets_bool)

=== Confusion Matrix Results ===
Accuracy: 0.6
Error Rate: 0.4
Sensitivity (Recall): 0.6666666666666666
Specificity: 0.5
Positive Predictive Value (Precision): 0.6666666666666666
Negative Predictive Value: 0.5
F-score: 0.6666666666666666

Confusion Matrix:
       Predicted
       Neg  Pos
Real Neg 1    1
Real Pos 1    2


In [16]:
function printConfusionMatrix(outputs::AbstractArray{<:Real,1}, targets::AbstractArray{Bool,1}; threshold::Real=0.5)
    # Call confusionMatrix function with threshold
    acc, errorRate, sens, spec, ppv, npv, fscore, confMat = confusionMatrix(outputs, targets, threshold=threshold)
    
    # Print results
    println("=== Confusion Matrix Results (Threshold: ", threshold, ") ===")
    println("Accuracy: ", acc)
    println("Error Rate: ", errorRate)
    println("Sensitivity (Recall): ", sens)
    println("Specificity: ", spec)
    println("Positive Predictive Value (Precision): ", ppv)
    println("Negative Predictive Value: ", npv)
    println("F-score: ", fscore)
    println("\nConfusion Matrix:")
    println("       Predicted")
    println("       Neg  Pos")
    println("Real Neg ", confMat[1,1], "    ", confMat[1,2])
    println("Real Pos ", confMat[2,1], "    ", confMat[2,2])
    println("================================")
end

printConfusionMatrix (generic function with 2 methods)

In [18]:
# Test second printConfusionMatrix (Real inputs)
println("\n--- Test with Real Inputs ---")
outputs_real = [0.1, 0.8, 0.3, 0.9, 0.2]
targets_real = [true, false, false, true, true]
printConfusionMatrix(outputs_real, targets_real)

println("\n--- Test with threshold 0.3 ---")
printConfusionMatrix(outputs_real, targets_real, threshold=0.3)


--- Test with Real Inputs ---
=== Confusion Matrix Results (Threshold: 0.5) ===
Accuracy: 0.4
Error Rate: 0.6
Sensitivity (Recall): 0.3333333333333333
Specificity: 0.5
Positive Predictive Value (Precision): 0.5
Negative Predictive Value: 0.3333333333333333
F-score: 0.4

Confusion Matrix:
       Predicted
       Neg  Pos
Real Neg 1    1
Real Pos 2    1

--- Test with threshold 0.3 ---
=== Confusion Matrix Results (Threshold: 0.3) ===
Accuracy: 0.2
Error Rate: 0.8
Sensitivity (Recall): 0.3333333333333333
Specificity: 0.0
Positive Predictive Value (Precision): 0.3333333333333333
Negative Predictive Value: 0.0
F-score: 0.3333333333333333

Confusion Matrix:
       Predicted
       Neg  Pos
Real Neg 0    2
Real Pos 2    1


The printConfusionMatrix functions displayed all calculated metrics in a clear, readable format. Testing with boolean inputs showed well-organized output with properly labeled metrics and a structured confusion matrix visualization. The real-valued input version correctly included the threshold parameter in the output header, making it easy to track which threshold was used for classification. Multiple tests with different thresholds (0.5 and 0.3) demonstrated that the print function consistently formats results, displaying how threshold changes affect all metrics simultaneously. The confusion matrix visualization with "Predicted" and "Real" labels provides an intuitive understanding of true positives, true negatives, false positives, and false negatives at a glance.